lib/logstash/inputs/redis.rb

```diff
@@ -217,6 +246,86 @@ def list_runner(output_queue)
     end
   end
 
+  # private
+  def list_pattern_stop
+    return if @redis.nil? || !@redis.connected?
+
+    @redis.quit rescue nil
+    @redis = nil
+    @threadpool.shutdown
+    @threadpool.wait_for_termination
+  end
+
+  # private
+  def worker_consume(output_queue, key)
+    @logger.debug("Launched worker for #{key}")
+    redis = new_redis_instance
+    begin
+      (0..@pattern_list_max_items-1).each do
+        if stop?
+          @logger.debug("Breaking from thread #{key} as it was requested to stop")
+          break
+        end
+        value = redis.lpop(key)
+        break if value.nil?
+        queue_event(value, output_queue)
+      end
+    ensure
+      redis.quit rescue nil
+    end
+    @logger.debug("Exit worker for #{key}")
+  end
+
+  # private
+  def threadpool_capacity?
+    @threadpool.remaining_capacity > 0
+  end
+
+  # private
+  def launch_worker(output_queue, key)
+    @current_workers.add(key)
+    @threadpool.post do
+      begin
+        worker_consume(output_queue, key)
+      ensure
+        @current_workers.delete(key)
+      end
+    end
+  end
+
+  # private
+  def ensure_workers(output_queue)
+    return unless threadpool_capacity?
+    keys = @redis.keys(@key)
+    keys.shuffle
```

Comment:
This doesn't seem to be doing anything as the non-bang (exclamation mark) version of this Ruby method doesn't mutate the object and instead returns a new object ( https://ruby-doc.org/core-2.4.1/Array.html#method-i-shuffle ). Instead of using the bang version, probably cleanest to chain this together as `@redis.keys(@key).shuffle.each do ...` so that there's least amount of mutations / assignments